src/modules/hip/kernel/spectrogram.hpp

```diff
@@ -0,0 +1,250 @@
+        for (int j = 0; j < hipBlockDim_x; j++)
+        {
+            realVal += (input_smem[hipThreadIdx_y][j] * cosFactor_smem[j][hipThreadIdx_x]);
+            imaginaryVal += (-input_smem[hipThreadIdx_y][j] * sinFactor_smem[j][hipThreadIdx_x]);
+        }
+        __syncthreads();
+    }
+
+    // final store to dst
+    if ((id_y < numWindows) && (id_x < numBins))
+    {
+        float magnitudeSquare = ((realVal * realVal) + (imaginaryVal * imaginaryVal));
+        
+        /* if vertical is set to true, then get the transposed output index (id_x * dstStrideNH.y + id_y)
+           else get the normal output index (id_y * dstStrideNH.y + id_x) */
+        int dstIdx = (vertical) ? (id_z * dstStrideNH.x + id_x * dstStrideNH.y + id_y) :
+                                  (id_z * dstStrideNH.x + id_y * dstStrideNH.y + id_x);
+        dstPtr[dstIdx] = (power == 2) ? magnitudeSquare : sqrtf(magnitudeSquare);
+    }
+}
+
+// -------------------- Set 1 - kernel executor --------------------
+
+RppStatus hip_exec_spectrogram_tensor(Rpp32f* srcPtr,
+                                      RpptDescPtr srcDescPtr,
+                                      Rpp32f* dstPtr,
+                                      RpptDescPtr dstDescPtr,
+                                      Rpp32s *srcLengthTensor,
+                                      bool centerWindows,
+                                      bool reflectPadding,
+                                      Rpp32f *windowFunction,
+                                      Rpp32s nfft,
+                                      Rpp32s power,
+                                      Rpp32s windowLength,
+                                      Rpp32s windowStep,
+                                      rpp::Handle& handle)
+{
+    bool vertical = (dstDescPtr->layout == RpptLayout::NFT);
+    Rpp32s numBins = (nfft / 2 + 1);
+    
+    // find the maximum windows required across all inputs in batch and stride required for window output
+    Rpp32s maxNumWindows = (vertical) ? dstDescPtr->w : dstDescPtr->h;
+    uint scratchMemorySize = nfft * ((maxNumWindows * dstDescPtr->n) + (numBins * 2));
+    
+    // check if scratch memory size required for spectrogram is within the limits
+    if (scratchMemorySize > SPECTROGRAM_MAX_SCRATCH_MEMORY)
+        return RPP_ERROR_OUT_OF_BOUND_SCRATCH_MEMORY_SIZE;
+
+    // generate hanning window
+    Rpp32f *windowFn;
+    if (windowFunction == NULL)
+    {
+        windowFn = handle.GetInitHandle()->mem.mcpu.scratchBufferHost;
+        hann_window(windowFn, windowLength);
+    }
+    else
+    {
+        windowFn = windowFunction;
+    }
+
+    // copy the hanning window values to hip memory
+    Rpp32f *d_windowFn = handle.GetInitHandle()->mem.mgpu.scratchBufferHip.floatmem;
+    CHECK_RETURN_STATUS(hipMemcpyAsync(d_windowFn, windowFn, windowLength * sizeof(Rpp32f), hipMemcpyHostToDevice, handle.GetStream()));
+    CHECK_RETURN_STATUS(hipStreamSynchronize(handle.GetStream()));
```

Comment:
Please remove `CHECK_RETURN_STATUS(hipStreamSynchronize(handle.GetStream())));` from here and line 198 and move it to after the last hipLaunchKernelGGL call.